src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java

```diff
@@ -0,0 +1,190 @@
+      return;
+    }
+
+    callSpan.finish();
+    spanInScope.close();
+
+    if (throwable != null) {
+      callSpan.annotate(throwable.toString());
+    }
+  }
+
+  @Override public void dnsStart(Call call, String domainName) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    dnsSpan =
+        tracer.newChild(callSpan.context()).start().name("dns");
+  }
+
+  @Override
+  public void dnsEnd(Call call, String domainName, @Nullable List<InetAddress> inetAddressList,
+      @Nullable Throwable throwable) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    if (throwable == null) {
+      dnsSpan.tag("dns results",
+          inetAddressList.stream().map(Object::toString).collect(Collectors.joining(", ")));
+    } else {
+      dnsSpan.annotate(throwable.toString());
+    }
+
+    dnsSpan.finish();
+  }
+
+  @Override public void connectStart(Call call, InetSocketAddress inetSocketAddress, Proxy proxy) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    connectSpan =
+        tracer.newChild(callSpan.context()).start().name("connect");
+  }
+
+  @Override public void connectEnd(Call call, InetSocketAddress inetSocketAddress,
+      @Nullable Protocol protocol, @Nullable Throwable throwable) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    if (throwable == null) {
+      connectSpan.tag("host", inetSocketAddress.toString());
+      connectSpan.tag("protocol", protocol.toString());
+    } else {
+      connectSpan.annotate(throwable.toString());
+    }
+
+    connectSpan.finish();
+
+    connectionEvent = true;
+  }
+
+  @Override public void connectionFound(Call call, Connection connection) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    if (!connectionEvent) {
+      tracer.newChild(callSpan.context())
+          .start()
+          .name("connect")
+          .finish();
+    }
+  }
+
+  @Override public void secureConnectStart(Call call) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    secureConnectSpan =
+        tracer.newChild(callSpan.context()).start().name("tls");
+  }
+
+  @Override public void secureConnectEnd(Call call, @Nullable Handshake handshake,
+      @Nullable Throwable throwable) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    if (throwable == null) {
+      secureConnectSpan.tag("handshake", handshake.toString());
+    } else {
+      secureConnectSpan.annotate(throwable.toString());
+    }
+
+    secureConnectSpan.finish();
+  }
+
+  @Override public void requestHeadersStart(Call call) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    requestSpan =
+        tracer.newChild(callSpan.context()).start().name("request");
+  }
+
+  @Override public void requestBodyEnd(Call call, Throwable throwable) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    requestSpan.finish();
+  }
+
+  @Override public void responseHeadersStart(Call call) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    responseSpan =
```

Comment:
so this is a span for reading the response. It probably makes sense to close the call span at this point as usually RPC span ends when headers are read (definitely no single rule here)